src/track/heaptrack_inject.cpp

```diff
@@ -176,9 +176,80 @@ struct dlopen
     static constexpr auto name = "dlopen";
     static constexpr auto original = &::dlopen;
 
+    static Dl_serinfo* load_serinfo(const char* module_path) noexcept
+    {
+        Dl_serinfo *serinfo = NULL;
+        Dl_serinfo serinfo_size;
+        void *module;
+
+        module = original(module_path, RTLD_LAZY);
+        if (module == NULL) {
+            return NULL;
+        }
+
+        int res = dlinfo(module, RTLD_DI_SERINFOSIZE, &serinfo_size);
+        if (res) {
+            dlclose(module);
+            return NULL;
+        }
+
+        serinfo = (Dl_serinfo*) ::malloc(serinfo_size.dls_size);
+        *serinfo = serinfo_size;
+        res = dlinfo(module, RTLD_DI_SERINFO, serinfo);
+        if (res) {
+            ::free(serinfo);
+            serinfo = NULL;
+        }
+        dlclose(module);
+        return serinfo;
+    }
+
     static void* hook(const char* filename, int flag) noexcept
     {
-        auto ret = original(filename, flag);
+        void *ret = NULL;
+        bool fallback = false;
+        if (filename == NULL || filename[0] == '/') {
+            // absolute path
+            ret = original(filename, flag);
+        }
+        if (!ret) {
+            // need to respect rpath
+            // get module info
+            Dl_info dl_info;
+            int res = dladdr(__builtin_return_address(0), &dl_info);
+            if (!res) {
+                fallback = true;
+            } else {
+                Dl_serinfo *serinfo;
```

Comment:
see above: here just do `Dl_serinfo serinfo;` and then pass `&serinfo` to the `load_serinfo` call below?